File: L03/figures/L03_amplitude_and_phase.m
%-------------------------------------------------------------------------------
% Function
%-------------------------------------------------------------------------------
function L03_amplitude_and_phase()

% clear command window
clc;

% parameters
fs = 1000;    % sampling rate of original signal
T = 1;        % signal duration
N = T * fs;   % number of samples

fc = 30;      % carrier frequency
f0 = 3;       % modulation frequency
M = 0.5;      % modulation factor (depth)

% time variable
t = linspace(0, T, N);

% AM signal
y = (1 + M * cos(2 * pi * t * f0));
u = sin(2 * pi * t * fc);
s = y .* u;

% signal amplitude (envelope) and phase
E = abs(hilbert(s));
P = angle(hilbert(s)) / (2 * pi);

% plot signals
subplot(2, 1, 1);
plot(t, s, 'k'); hold on;
% plot(t, y, 'r');
set(gca, 'XLim', [t(1), t(end)], 'YLim', [-1.6, 1.6], 'FontSize', 12);
set(get(gca, 'XLabel'), 'String', 'time (s)', 'FontSize', 12);
set(get(gca, 'YLabel'), 'String', 'amplitude', 'FontSize', 12);

subplot(2, 1, 2);
plot(t, E, 'r'); hold on;
plot(t, P, 'b');
set(gca, 'XLim', [t(1), t(end)], 'YLim', [-1.6, 1.6], 'FontSize', 12);
set(get(gca, 'XLabel'), 'String', 'time (s)', 'FontSize', 12);
set(get(gca, 'YLabel'), 'String', 'amplitude', 'FontSize', 12);

end % end 

%-------------------------------------------------------------------------------

File: L03/codes/L03_amplitude_and_phase.py
# -*- coding: utf-8 -*-

import numpy as np
import matplotlib.pyplot as plt

from scipy import signal

#------------------------------------------------------------------------------
# Function, main()
#------------------------------------------------------------------------------
def L03_main():
    
    # parameters
    fs = 1000   # sampling rate, in Hz
    N  = 1000   # duration, in samples
    T  = N / fs # duration, in seconds
    
    f0 = 3      # signal frequency, in Hz
    fc = 30     # carrier frequency, in Hz
    AM = 0.5    # modulation factor
    
    # time variable
    t = np.linspace(0, T, N)
    
    # amplitude modulated signal
    y = 1 + AM * np.cos(2 * np.pi * f0 * t)
    s = np.sin(2 * np.pi * fc * t) * y
    
    # detect amplitude and phase
    A = np.abs(signal.hilbert(s))
    P = np.angle(signal.hilbert(s)) / (2 * np.pi)
    
    # plot
    plt.subplot(2, 1, 1)
    plt.plot(t, s)
    plt.plot(t, y, linestyle='--')
    plt.xlim(0, T)
    plt.ylim(-2.0, 2.0)
    
    plt.subplot(2, 1, 2)
    plt.plot(t, P)
    plt.plot(t, A, linestyle='--')
    plt.xlim(0, T)
    plt.ylim(-2.0, 2.0)
    
#------------------------------------------------------------------------------
if __name__ == '__main__':
    L03_main()
